File: 785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        
        int[] colors = new int[n+1];
        colors[n] = -1;
        
        for(int i = 0; i < n; i++){
            if(colors[i] == 0 && !dfs(i,n,colors,graph)) return false;    
        }
        
        return true;
    }
    
    private boolean dfs(int n,int p, int[] colors, int[][] graph){
        if(colors[n] != 0) return colors[p] != colors[n];

        colors[n] = -1 * colors[p]; // mark visited and also color the node
        for(int adj : graph[n])
            if(!dfs(adj,n,colors,graph)) return false;
        return true;
    }
}

File: 785-is-graph-bipartite/785-is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        
        n = len(graph)
        visited = [0] * n
        '''
            visited[i] = -1 => visited and colored white 
            visited[i] = 0 => not_visited 
            visited[i] = 1 => visited and colored black
        '''
        
        def dfs(i,p):
            if(visited[i] == p):
                return False
            elif(visited[i] == -1*p):
                return True
            visited[i] = -1*p
            for adj in graph[i]:
                if(not dfs(adj,-1*p)):
                    return False
            return True
        
        for i in range(n):
            if(visited[i] == 0 and not dfs(i,-1)):
                return False
            
        return True
                